### PR TITLE
[docs]: Add description for types

### PR DESCRIPTION
**Pull request by erbesharat**
**What this PR does / why we need it**:
Adds description for types

**Special notes for your reviewer**:
* Do we need to document experimental types too?
* Let me know if there are any grammar issues or typo, or if I've missed something or something needs more clarification.

### CODE REVIEW

File: content/docs/specification/content.md [hunk 1]
```diff
@@ -43,151 +43,193 @@ Further links about URL encoding:
...
-**Type**  
 
-* Key: **type**
-* Mandatory
-* Permitted values: "host"
-* Example: "type=host"
+## Description
 
-**Version**  
+`host` is our most simple type. Using `host` you can redirect an incoming request to a specific endpoint with a custom status code defined in the TXT record.
```
Comment (stp-ip) [inline content/docs/specification/content.md]:
```suggestion
`host` is the most simple type. Using `host` you can redirect an incoming request to a specific endpoint with a custom status code defined in the TXT record.
```

File: content/docs/specification/content.md [hunk 2]
```diff
@@ -43,151 +43,193 @@ Further links about URL encoding:
...
+For example records you can check [host](https://about.txtdirect.org/docs/examples/#host) in the examples section.
```
Comment (stp-ip) [inline content/docs/specification/content.md]:
```suggestion
For example records you can check [host](https://about.txtdirect.org/docs/examples/#host).
```

Simpler and I think self explanatory.

File: content/docs/specification/content.md [hunk 3]
```diff
@@ -43,151 +43,193 @@ Further links about URL encoding:
...
-**Type**  
 
-* Key: **type**
-* Mandatory
-* Permitted values: "path"
-* Example: "type=path"
+## Description
 
-**Version**  
+Sometimes you may need to redirect a request separate endpoints based on the request's path content. `path` type runs a regular expression which can be our default regex or a custom regex from the TXT record. Then uses the matches from that expression to generate a zone address and find final record to use it for redirecting the request.
```
Comment (stp-ip) [inline content/docs/specification/content.md]:
```suggestion
Sometimes you may need to redirect a request coming into a single host/domain to separate locations based on the request's path. One common use-case is a URL shortener.
`path` type extracts each directory by default. Each directory is use to generate a new sub zone to fetch the TXT record from. To reorder the default order you can use the `from=` field.
```

File: content/docs/specification/content.md [hunk 4]
```diff
@@ -43,151 +43,193 @@ Further links about URL encoding:
...
 
-* Key: **v**
-* Mandatory
-* Permitted values: "txtv0"
-* Example: "v=txtv0"
+`path` also contains some additional configuration to process the request's path. For example by using `from=` field inside a `path` record you can define how we should order the matches from the expression on the request's path to generate the next zone address. Or add the `re=` field to use a custom regex instead of our default on the request's path.
```
Comment (stp-ip) [inline content/docs/specification/content.md]:
```suggestion
`path` also contains some additional configuration to process the request's path. The `re=` field enables use of a custom regex to control which data is used to construct the sub zones, where TXT record data is fetched.
```

File: content/docs/specification/content.md [hunk 5]
```diff
@@ -43,151 +43,193 @@ Further links about URL encoding:
...
-**Redirect URL**  
+### Path Sanitization
 
-* Key: **to**
-* Recommended
-* Default: Fallback to **www** or **redirect** config
-* Permitted values: "absolute|relative URL"
-* Example: "to=https://example.com"
-* Example: "to=/example/"
+To use the request's path for generating the zone address we need normalize the path into allowed DNS chars. For example we need to replace `.`(dot) into `-`(dash) to keep each match in its own zone. You can visit [RFC1034](https://tools.ietf.org/html/rfc1034) for more information about the allowed chars.
```
Comment (stp-ip) [inline content/docs/specification/content.md]:
```suggestion
To use the request's path for generating the zone address it needs to be normalized into allowed DNS chars. For example `.`(dot) is replaced with `-`(dash) to keep each match in its own zone. You can visit [RFC1034](https://tools.ietf.org/html/rfc1034) for more information about the allowed chars.
```

Is this the only normalization that we do? `.` -> `-`?

File: content/docs/specification/content.md [hunk 6]
```diff
@@ -43,151 +43,193 @@ Further links about URL encoding:
...
+Sometimes you may need to redirect a request separate endpoints based on the request's path content. `path` type runs a regular expression which can be our default regex or a custom regex from the TXT record. Then uses the matches from that expression to generate a zone address and find final record to use it for redirecting the request.
 
-* Key: **v**
-* Mandatory
-* Permitted values: "txtv0"
-* Example: "v=txtv0"
+`path` also contains some additional configuration to process the request's path. For example by using `from=` field inside a `path` record you can define how we should order the matches from the expression on the request's path to generate the next zone address. Or add the `re=` field to use a custom regex instead of our default on the request's path.
 
-**Redirect URL**  
+### Path Sanitization
```
Comment (stp-ip) [inline content/docs/specification/content.md:99]: Is it path or zone normalization? Not sure, which one makes it clearer.

File: content/docs/specification/content.md [hunk 7]
```diff
@@ -43,151 +43,193 @@ Further links about URL encoding:
...
-* Key: **to**
-* Recommended
-* Default: Fallback to **www** or **redirect** config
-* Permitted values: "absolute|relative URL"
-* Example: "to=https://example.com"
-* Example: "to=/example/"
+To use the request's path for generating the zone address we need normalize the path into allowed DNS chars. For example we need to replace `.`(dot) into `-`(dash) to keep each match in its own zone. You can visit [RFC1034](https://tools.ietf.org/html/rfc1034) for more information about the allowed chars.
 
-**Root Redirect URL**  
+After this step we run our default regex or the custom regex from `re=` field on the request's path. If `from=` is not available we first revert order of matches from the expression then join them with request's host to generate the next zone in the chain. But if `from=` is availabe, we use the given order instead of reverting.
```
Comment (stp-ip) [inline content/docs/specification/content.md]:
```suggestion
The request's path is used to construct the zone address, which provides the redirect data used for this specific path. By default each directory is extracted and the order reverted meaning the less specific zone will come first:
`example.com/firstDirectory/secondDirectory/ -> secondDirectory.firstDirectory.example.com`
A custom ordering can be achieved by using the `from=` field. It will reorder the sub zones accordingly.

If a custom regex is configured with the `re=` field each match is used as a sub zone. Ordering can be modified by using named matches. The default ordering will use the first match as the first sub zone and continue until there are no new matches.
```

Not sure if that makes it clearer. Happy for other iterations.

File: content/docs/specification/content.md (hunk 4 above)
Comment (stp-ip) [reply content/docs/specification/content.md]: Docs are usually generalize from a PoV so it's not we, but it or passive voice

File: content/docs/specification/content.md (hunk 7 above)
Comment (stp-ip) [reply content/docs/specification/content.md]: Referencing previous paragraphs or steps is harder to follow in my view. Each section should mostly stand on it's own are at least not use text referencing a previous item. Harder to take out of context then.

File: content/docs/specification/content.md [hunk 8]
```diff
@@ -43,151 +43,193 @@ Further links about URL encoding:
...
+- Default: Fallback to **www** or **redirect** config
+- Permitted values: "absolute|relative URL"
+- Example: "to=https://example.com"
+- Example: "to=/example/"
+
+**Simplified Regex**
+
+- Key: **from**
+- Permitted values: "simplified regex"
+- Example: "from=/$2/$1/\$3"
```
Comment (stp-ip) [inline content/docs/specification/content.md]: Why is there an `\` added?

File: content/docs/specification/content.md (hunk 2 above)
Comment (mpourismaiel) [inline content/docs/specification/content.md]: For example you can checkout sample records in the [host](...) section in our examples.

File: content/docs/specification/content.md [hunk 9]
```diff
@@ -43,151 +43,193 @@ Further links about URL encoding:
...
-**Type**  
 
-* Key: **type**
-* Mandatory
-* Permitted values: "dockerv2"
-* Example: "type=dockerv2"
+## Description
 
-**Version**  
+To redirect request from Docker CLI to a specific docker image or a docker registry you can use `dockerv2` type. The difference between a Docker CLI request and a normal request is that Docker CLI sends a couple of request with different paths to fetch some information like image tags, manifests and etc. about the image. To handle these requests we use docker's v2 API to fetch the data for each request.
```
Comment (stp-ip) [inline content/docs/specification/content.md]: Not sure about redirecting. That's done in the backend. For the user it's probably more about controlling the endpoint or have a vanity url. How about:
```suggestion
The `dockerv2` type enables vanity URLs for your container images. Using your own URl for container images enables easier switching between your backend infrastructure/storage and adds more control.
The type implements the Docker registry API v2 and redirects requests for both metadata and image blobs to the backend storage. All data besides the redirect is therefore served by your underlying infrastructure such as Google Container Registry (gcr.io).
```

Probably a general issue is that we are conflating the specification a bit with user docs. Ok for now but in the long term with a new implementation we might need to switch the language from user specific to implementation specific.

So instead of "can be used for" to "the type needs to implement the docker registry api v2 as documented here".

File: content/docs/specification/content.md (hunk 3 above)
Comment (mpourismaiel) [inline content/docs/specification/content.md]: redirect a request **to** separate endpoints.

File: content/docs/specification/content.md (hunk 5 above)
Comment (mpourismaiel) [inline content/docs/specification/content.md]: we need **to** normalize...path into _valid_ DNS chars.

File: content/docs/specification/content.md (hunk 9 above)
Comment (mpourismaiel) [inline content/docs/specification/content.md]: sends a couple of _requests_ with ... manifests and etc. _regarding_ the image.

File: content/docs/specification/content.md [hunk 10]
```diff
@@ -43,151 +43,193 @@ Further links about URL encoding:
...
 
-* Key: **v**
-* Mandatory
-* Permitted values: "txtv0"
-* Example: "v=txtv0"
+You can point to a docker registry or a specific image inside your record. With pointing to a registry you can have vanity urls for all the images on that registry instead of only an image.
```
Comment (stp-ip) [inline content/docs/specification/content.md]: This needs to be made a bit clearer. Not happy with my solution either, but thought I provide a few thoughts and you can iterate on what you think might be a good solution.

```suggestion
The upstream endpoint used for `to=` can be different depending on the use case.
1. Image link including a specific tag enables serving this specific tag under various image names and tags.
2. Image link enables serving available tags for a specific image under various image names.
3. Referencing the root of a container registry enables the full serving of the registry under this specific domain.
```

File: content/docs/specification/content.md [hunk 11]
```diff
@@ -43,151 +43,193 @@ Further links about URL encoding:
...
+
+**Website Redirect URL**
+
+- Key: **website**
+- Recommended
+- Default: Fallback to **www** or **redirect** config
+- Note: "Non container traffic redirect"
+- Permitted values: "absolute|relative URL"
+- Example: "to=https://about.txtdirect.org/docs/"
+- Example: "to=/docs/"
```
Comment (stp-ip) [inline content/docs/specification/content.md]: In general does a relative URL make sense here? Wouldn't that fall back again?

File: content/docs/specification/content.md [hunk 12]
```diff
@@ -198,60 +240,63 @@ https://tools.ietf.org/html/rfc3986#page-11
...
-* Permitted values: "absolute URL"
-* Example: "to=https://example.com/subpage/"
-<!--root/website?-->
+**Upstream URL**
+
+- Key: **to**
+- Mandatory
+- Permitted values: "absolute URL"
+- Example: "to=https://example.com/subpage/"
+  <!--root/website?-->
```
Comment (stp-ip) [inline content/docs/specification/content.md]: Do we need a root/website? That was still an open question, but I think with a proxy we would most likely not need this as all failures should be handled by the upstream. If you agree, we can remove this comment.

**Review by mpourismaiel — COMMENTED** [review]
Just a friendly note :D

File: content/docs/specification/content.md [hunk 13]
```diff
@@ -43,151 +43,193 @@ Further links about URL encoding:
...
-**Type**  
 
-* Key: **type**
-* Mandatory
-* Permitted values: "gometa"
-* Example: "type=gometa"
+## Description
 
-**Version**  
+One of the features that `go get` command has is that you can point a webpage to a Go package using go-import meta tag. For more information about these tags please visit [Remote import paths](https://golang.org/cmd/go/#hdr-Remote_import_paths). Using `gometa` type you can point to a go package inside your records using `to=` field and we use that URI to generate a simple HTML page that only contains go-import and go-source **(Only for packages served on GitHub)** tags for `go get` to use it and find your package.
```
Comment (stp-ip) [inline content/docs/specification/content.md]: This is explaining too much about go get and the go-import meta tag.
I would say something along the lines of: vanity url, control your import paths independent of your backend code repo (github/gitlab). easily switchable without impacting downstream users of your packages.
Using go-import meta tag under to implement this.

This part is great:
```
Using `gometa` type you can point to a go package inside your records using `to=` field and we use that URI to generate a simple HTML page that only contains go-import and go-source **(Only for packages served on GitHub)** tags for `go get` to use it and find your package.
```

**Review by stp-ip — CHANGES_REQUESTED** [review]
Great additions. Added a few more suggestions. Don't worry about the suggestions. I was more elaborate on the iterations to give feedback.
Mostly small restructures and moving from we to a more passive style.

Also experimental types can just have the specification for now and documenting them as soon as we feel they are stable. We might change how we use them so documentation might be out of date. What do you think?

We don't have our gomod one specified. Doesn't need a description, but should probably mention the record fields for now.

Thanks for taking a stab at the docs.

**Comment by stp-ip** [comment]
Hehe what's happening. Collaboration on docs. I'm so happy \o/

File: content/docs/specification/content.md (hunk 5 above)
Comment (erbesharat) [reply content/docs/specification/content.md]: I think `.`->`-` is the only one. Or maybe I couldn't find anything else.

File: content/docs/specification/content.md (hunk 6 above)
Comment (erbesharat) [reply content/docs/specification/content.md:99]: Since normalization happens on the path and its result is the zone I think *Path* seems better.

File: content/docs/specification/content.md (hunk 8 above)
Comment (erbesharat) [reply content/docs/specification/content.md]: It's VScode's fault :sweat_smile:

File: content/docs/specification/content.md (hunk 5 above)
Comment (stp-ip) [reply content/docs/specification/content.md]: How about special characters. We should probably test various special characters such as emojiis and unicode chars?

Comment (erbesharat) [reply content/docs/specification/content.md]: Currently, our default path regex ignores special chars and emojis. Should we replace them with dash when a custom regex is being used?

File: content/docs/specification/content.md (hunk 11 above)
Comment (erbesharat) [reply content/docs/specification/content.md]: I think it falls back too, doesn't make sense to use relative URLs for fallback endpoint.

File: content/docs/specification/content.md (hunk 5 above)
Comment (stp-ip) [reply content/docs/specification/content.md]: Let's create a separate issue to test and handle these edge cases. Maybe interesting would be testing with a fuzzer and see, if we actually handle all edge cases without us coding up manual tests.

File: content/docs/specification/content.md (hunk 11 above)
Comment (stp-ip) [reply content/docs/specification/content.md]: Ok let's remove this from the docs at least.

**Comment by stp-ip** [comment]
A few issues are still open or Github is having issues showing them as resolved. Let me know, when it's ready for another review.

**Comment by erbesharat** [comment]
@stp-ip It's ready for another review.

**Review by stp-ip — APPROVED** [review]
